lib/msal-core/src/cache/AuthCache.ts

```diff
@@ -87,19 +88,29 @@ export class AuthCache extends BrowserStorage {// Singleton
      * @param homeAccountIdentifier 
      * @param tokenType 
      */
-    private matchKeyForType(key:string, clientId: string, homeAccountIdentifier: string, tokenType: string): Boolean {
-        const accountMatch = key.match(clientId) && key.match(homeAccountIdentifier);
-        
+    private matchKeyForType(key:string, clientId: string, homeAccountIdentifier: string, tokenType: string): AccessTokenKey {
+        // All valid token cache item keys are valid JSON objects, ignore keys that aren't
+        if (!StringUtils.isValidJson(key)) {
+            return null;
+        }
+
+        // Does the cache item match the request account
+        const accountMatches = key.match(clientId) && key.match(homeAccountIdentifier);
+        // Does the cache item match the requested token type
+        let tokenTypeMatches = false;
+
         switch (tokenType) {
             case ServerHashParamKeys.ACCESS_TOKEN:
                 // Cache item is an access token if scopes are included in the cache item key
-                return !!(accountMatch && key.match(Constants.scopes));
+                tokenTypeMatches = !!key.match(Constants.scopes);
+                break;
             case ServerHashParamKeys.ID_TOKEN:
-                // Cache item is an ID token if scopes are NOT included in the cache item key
-                return !!(accountMatch && !key.match(Constants.scopes));
-            default:
-                return false;
+                // Cache may be an ID token if scopes are NOT included in the cache item key
+                tokenTypeMatches = !key.match(Constants.scopes);
+                break;
         }
+
+        return (accountMatches && tokenTypeMatches) ? JSON.parse(key) : null;
```

Comment:
FYI: This `JSON.parse doesn't` need to be inside a `try/catch` block because the first conditional check in the method has already ruled out an invalid JSON key. 

If anyone thinks we should wrap this line in a `try/catch` as a sanity check, I can add it.

Reply:
Could the `isValidJson` function return the parsed object?

Reply:
+1

Reply:
Done, the method seems a little awkward to me, open to suggestions on how to simplify/clean it up.

Reply:
I like it!